packages/camera/camera_windows/windows/capture_controller.cpp

```diff
@@ -0,0 +1,1185 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "capture_controller.h"
+
+#include <wincodec.h>
+#include <wrl/client.h>
+
+#include <cassert>
+#include <system_error>
+
+#include "string_utils.h"
+
+namespace camera_windows {
+
+using Microsoft::WRL::ComPtr;
+
+CaptureControllerImpl::CaptureControllerImpl(
+    CaptureControllerListener *listener)
+    : capture_controller_listener_(listener), CaptureController(){};
+
+CaptureControllerImpl::~CaptureControllerImpl() {
+  ResetCaptureController();
+  capture_controller_listener_ = nullptr;
+};
+
+// static
+bool CaptureControllerImpl::EnumerateVideoCaptureDeviceSources(
+    IMFActivate ***devices, UINT32 *count) {
+  ComPtr<IMFAttributes> attributes;
+
+  HRESULT hr = MFCreateAttributes(&attributes, 1);
+
+  if (SUCCEEDED(hr)) {
+    hr = attributes->SetGUID(MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+                             MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE_VIDCAP_GUID);
+  }
+
+  if (SUCCEEDED(hr)) {
+    hr = MFEnumDeviceSources(attributes.Get(), devices, count);
+  }
```

Comment:
This is a weird pattern. Why not:
```
if (!SUCCEEDED(hr)) {
  return hr;
}
```

so that all of the normal logic is the main body, instead of appearing to be conditional?

In addition to being odd to read; it seems dangerous; it would be very easy to accidentally have one stray line of code later that wasn't gated and ran even for failures.

Reply:
You are right, this pattern is left over from goto pattern which was used on original MediaFoundation documentation. After I moved to smart pointers (ComPtr) these should have been rewritten as well.
I'll fix these.